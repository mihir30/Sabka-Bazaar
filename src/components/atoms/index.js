import Button from "./Button/Button";
import Image from "./Image/Image";
import WithSpinner from "./Spinner/Spinner";

export { Button, Image, WithSpinner };
