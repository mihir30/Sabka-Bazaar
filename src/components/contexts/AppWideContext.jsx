import React, { createContext, useReducer } from "react";
import { reducer } from "../reducers/ReducerContext";

const initialState = {
  count: 0,
  products: {},
};

export const AppWideContext = createContext(initialState);

const AppWideContextProvider = ({ children }) => {
  const [cartItems, dispatch] = useReducer(reducer, initialState);

  return (
    <AppWideContext.Provider value={{ cartItems, dispatch }}>
      {children}
    </AppWideContext.Provider>
  );
};

export default AppWideContextProvider;
