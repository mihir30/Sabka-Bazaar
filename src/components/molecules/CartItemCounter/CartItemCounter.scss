$color-white: white;
$background-color-pink: #cc0044;

.item-quantity {
    padding: 15px;
    font-weight: bold;
}

.item-price {
    font-weight: bold;
}

.item-price-multiply {
    margin: 7px;
    font-weight: bold;
}

.item-total-price {
    margin-left: 40px;
    font-weight: bold;
}

.btn-cart-qty {
    background-color: $background-color-pink;
    width: 30px;
    border-radius: 5px;
    margin: 15px 0 0 0;
    color: $color-white;
    border: 2px solid #cc0044;
    font-size: 18px;
}

@media only screen and (min-width: 1024px) {
    .btn-cart-qty {
        background-color: $background-color-pink;
        width: 30px;
        border-radius: 50%;
        margin: 15px 0 0 0;
        color: $color-white;
        border: 2px solid #cc0044;
        font-size: 18px;
    }
}