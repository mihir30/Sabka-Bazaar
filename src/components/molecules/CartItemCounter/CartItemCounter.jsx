import React, { useContext } from "react";
import { Button } from "../../atoms";
import { AppWideContext } from "../../contexts/AppWideContext";
import "./CartItemCounter.scss";
const CartItemCounter = ({ quantity, price, id, stock }) => {
  const { dispatch } = useContext(AppWideContext);

  const editItem = () => {
    dispatch({
      type: "EDIT_ITEM",
      id: id,
    });
  };

  const removeItem = () => {
    dispatch({
      type: "REMOVE_ITEM",
      id: id,
    });
  };

  const totalItemPrice = quantity * price;
  return (
    <div>
      <Button
        buttonType={"cart-qty"}
        children={"-"}
        type={"button"}
        method={() => removeItem()}
      />
      <span className="item-quantity" data-testid="item-count">{quantity}</span>
      <Button
        buttonType={"cart-qty"}
        children={"+"}
        type={"button"}
        method={() => editItem()}
        disabled={stock === quantity}
      />
      <span className="item-price-multiply" data-testid="item-multiply">X </span>
      <span className="item-price" data-testid="price">{`Rs.${price}`}</span>
      <span className="item-total-price" data-testid="total-price"> {`Rs.${totalItemPrice}`}</span>
    </div>
  );
};

export default CartItemCounter;
