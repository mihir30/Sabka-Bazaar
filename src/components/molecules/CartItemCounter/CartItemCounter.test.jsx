import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { AppWideContext } from '../../contexts/AppWideContext';
import { Button } from '../../atoms';
import CartItemCounter from './CartItemCounter';

describe("Cart Item Counter Component", () => {
    it('renders the component properly and should display the proper initial count in the cart item counter component', () => {
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={1} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("item-count")).toHaveTextContent(1);
    });

    it('should increment when + is clicked', () => {
        const editItem = jest.fn();
        render(<Button
            buttonType={"cart-qty"}
            children={"+"}
            type={"button"}
            method={() => editItem()}
            disabled={50 === 2}
          />);
        const buttonElement = screen.getByRole("button", { name: "+" });
        fireEvent.click(buttonElement);
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={2} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("item-count")).toHaveTextContent(2);
    });

    it('should decrement when - is clicked', () => {
        const removeItem = jest.fn();
        render(<Button
            buttonType={"cart-qty"}
            children={"-"}
            type={"button"}
            method={() => removeItem()}
          />);
        const buttonElement = screen.getByRole("button", { name: "-" });
        fireEvent.click(buttonElement);
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={1} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("item-count")).toHaveTextContent(1);
    });

    it('should display X', () => {
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={1} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("item-multiply")).toHaveTextContent("X");
    });

    it('should display price of product added to cart', () => {
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={1} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("price")).toHaveTextContent("187");

    });

    it('should display total item price', () => {
        render(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={3} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>);
        expect(screen.getByTestId("total-price")).toHaveTextContent("561");
    })

    it('should have the right cart item component snapshot', () => {
        const tree = renderer.create(<AppWideContext.Provider value={dispatchEvent}><BrowserRouter><CartItemCounter quantity={1} price={187} stock={50} id={"5b6c6aeb01a7c38429530884"}/></BrowserRouter></AppWideContext.Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

