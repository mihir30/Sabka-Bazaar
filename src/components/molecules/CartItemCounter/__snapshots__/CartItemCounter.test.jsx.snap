// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Item Counter Component should have the right cart item component snapshot 1`] = `
<div>
  <button
    className="btn-cart-qty"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <span
    className="item-quantity"
    data-testid="item-count"
  >
    1
  </span>
  <button
    className="btn-cart-qty"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <span
    className="item-price-multiply"
    data-testid="item-multiply"
  >
    X 
  </span>
  <span
    className="item-price"
    data-testid="price"
  >
    Rs.187
  </span>
  <span
    className="item-total-price"
    data-testid="total-price"
  >
     
    Rs.187
  </span>
</div>
`;
